src/transformers/trainer.py

```diff
@@ -2382,14 +2382,22 @@ def _save_checkpoint(self, model, trial, metrics=None):
         # Place checkpoint in final location after all saving is finished.
         # First wait for everyone to finish writing
         self.args.distributed_state.wait_for_everyone()
-        # Then go through the rewriting process starting on process 0
+        # Then go through the rewriting process, only renaming from main process(es)
         if staging_output_dir != output_dir:
-            with self.args.main_process_first(
-                desc="Renaming model checkpoint folder to true location", local=self.args.save_on_each_node
+            if (
+                self.args.distributed_state.is_local_main_process
+                if self.args.save_on_each_node
+                else self.args.distributed_state.is_main_process
             ):
                 if os.path.exists(staging_output_dir):
                     os.rename(staging_output_dir, output_dir)
 
+                    # Ensure rename completed in cases where os.rename is not atomic
+                    fd = os.open(output_dir, os.O_RDONLY)
+                    os.fsync(fd)
+
+            self.args.distributed_state.wait_for_everyone()
+
         # Maybe delete some older checkpoints.
         if self.args.should_save:
             self._rotate_checkpoints(use_mtime=True, output_dir=run_dir)
```

Comment:
Let's also move this to be under this if/else so that way we can have it be done just on a single process as needed

Reply:
(Above the `wait_for_everyone`)

Reply:
I actually have a related question. I noticed that the current code has push_to_hub unguarded by any check for is_local / is_world zero. This seems incorrect. However, I don't use that option so I didn't want to touch it without understanding implications.

My guess is that it would have been best to do push_to_hub after wait_for_everyone from the final output_dir. Otherwise it seems like the push could end up shipping partially written state.

Reply:
I reworked the logic for this so that we include the rotate_checkpoints function. Also curious about push_to_hub and the save_to_json that is above line 2384.

Reply:
Agreed thinking more on it, it should be after for the reasons you mentioned

Reply:
Also, `push_to_hub` checks `is_world_process_zero()`, so it's fine :)